### ISSUE TITLE
C++ Header-Only Libraries: Dependencies

### ISSUE THREAD

**Issue by ax3l** [spack/spack#2378]
One question I forgot to ask at SC16 is, how to handle `depends_on` of C++ header-only libraries.

A C++ header-only library can, e.g., introduce a dependency to `boost` (can again be header only but additionally also link-time). Currently, `depends_on(type=...)` "build" does not forward the include paths:
  *made available during the project’s build. The package will be added to PATH, the compiler include paths[...]. Other projects which depend on this one **will not have these modified** [...]* ([ref](http://spack.readthedocs.io/en/latest/packaging_guide.html#dependency-specs))

but the problem is, that a header-only library's build time does nothing and it's install only copies it's files.

The type  "link" only adds the libraries (but not the includes). The last available option "run" does not apply here.

How to handle such libraries without forcing the user of those to look up all sub (sub-sub, sub-sub-sub) dependencies? Do we need a new "depends_on" type?

**Comment by tgamblin** [comment]
`link` should work for this, though it will add an extra `RPATH` that we could probably get rid of.  A `link` dependency gets the headers and the includes forwarded.

Where's the text above from in the docs?

**Comment by ax3l** [comment]
Oh that sounds great! The "link" [docs currently only state](http://spack.readthedocs.io/en/latest/packaging_guide.html#dependencies):
*the project is linked to by the project. The package will be added to the current package’s rpath.*

**Comment by citibeth** [comment]
Look for `google-test` for examples.

**Comment by ax3l** [comment]
thank you, great example for a missing "make install" but does not have any dependencies declared.

**Comment by adamjstewart** [comment]
@scheibelp

**Comment by ax3l** [comment]
@alalazo I often have the case of installing C++ header-only libraries that pull in dependencies.
The install as described in this thread works well, but if I want to "use" those libraries with modules, similar to the debian `...-dev` packages, I need to load the module *and all it's dependencies* manually, so `CMAKE_PREFIX_PATH` is set for e.g. header files.

Is there a way in the [Package File API](https://spack.readthedocs.io/en/latest/module_file_support.html#package-file-api) to already express: "also load the dependent modules" (because there is no simple R-Path that can be chained to the indirect dependencies)?

Basically, for C++ header only libraries I would love to express the module `autoload` directly in the `Package API` and not as late as in the `modules.yaml`, because otherwise I would need to maintain two places for such packages.

**Comment by ax3l** [comment]
@alalazo any idea? :)

**Comment by alalazo** [comment]
> Basically, for C++ header only libraries I would love to express the module autoload (preferably over prereq) directly in the Package API and not as late as in the modules.yaml, because otherwise I would need to maintain two places for such packages (or expose all indirect dependencies to users so they can module load them themselves

Currently there's nothing like that. The nearest thing you can do is construct directly the modifications:
```python
# Pseudo-code, so there may be errors
def setup_environment(self, build_env, run_env):
    ld_library_path = []
    for x in self.spec.traverse():
        ld_library_path.append(x.prefix.lib)

    run_env.append_path('LD_LIBRARY_PATH', ':'.join(ld_library_path))
```

I'll add this request to the list of improvements I need to implement for module file generation and either push it to #3183 or to a later PR (if that one goes in in the meanwhile).

**Comment by ax3l** [comment]
actually an additional `depends_on` `type` such as `"forward"` would be neat, too.

Once could allow header-only libraries with dependencies just to recklessly inject a `depends_on` in a dependent, e.g. for work-arounds as in #5014

**Comment by ax3l** [comment]
I think that using Spack environments solves this problem these days :)
